osh/sh_expr_eval.py

```diff
@@ -386,57 +385,63 @@ def _StringToBigInt(self, s, blame_loc):
-                    integer = self.EvalToBigInt(node2)
-                else:
-                    # BoolEvaluator doesn't have parse_ctx or mutable_opts
-                    assert self.mutable_opts is not None
+            e_die("Invalid integer constant %r" % s, blame_loc)
 
-                    # We don't need to flip _allow_process_sub, because they can't be
-                    # parsed.  See spec/bugs.test.sh.
-                    with state.ctx_Option(self.mutable_opts,
-                                          [option_i._allow_command_sub],
-                                          False):
-                        integer = self.EvalToBigInt(node2)
+        return integer
 
-            else:
-                if len(s.strip()) == 0 or match.IsValidVarName(s):
-                    # x42 could evaluate to 0
-                    e_strict("Invalid integer constant %r" % s, blame_loc)
-                else:
-                    # 42x is always fatal!
-                    e_die("Invalid integer constant %r" % s, blame_loc)
+    def _StringToBigInt(self, s, blame_loc):
+        # type: (str, loc_t) -> mops.BigInt
+        """Dynamically parse and evaluate a bash arith expression.
+
+        Runtime parsing enables silly stuff like
+
+            $(( $(echo 1)$(echo 2) + 1 )) => 13
+
+        This is called recursively. So a substitution such as
+
+            $(( "1 + 0x2 + 03" ))
+
+        in turn calls:
+          _StringToBigInt('1 + 0x2 + 03')
+            > _StringToBigInt('1')
+            > _StringToBigInt('0x2')
+            > _StringToBigInt('03')
+        """
+        # True when evaluating expressions inside 'test' and '['. They only
+        # expect an integer literal, not an arbitrary expression.
+        if self.parse_ctx is None:
+            return self._ParseInt(s, blame_loc)
+
+        # Special case so we don't get EOF error
+        if len(s.strip()) == 0:
+            return mops.ZERO
+
+        # Dynamically parse for compatibility
+        a_parser = self.parse_ctx.MakeArithParser(s)
+        try:
+            node2 = a_parser.Parse()  # may raise error.Parse
+        except error.Parse as e:
+            self.errfmt.PrettyPrintError(e)
+            e_die('Parse error in recursive arithmetic', e.location)
+
+        # This is just a single word, try to parse it as an integer literal
+        if node2.tag() == arith_expr_e.Word:
+            return self._ParseInt(s, blame_loc)
+
+        if self.exec_opts.eval_unsafe_arith():
+            integer = self.EvalToBigInt(node2)
+        else:
+            # BoolEvaluator doesn't have parse_ctx or mutable_opts
+            assert self.mutable_opts is not None
+
+            # We don't need to flip _allow_process_sub, because they can't be
+            # parsed.  See spec/bugs.test.sh.
+            with state.ctx_Option(self.mutable_opts,
+                                  [option_i._allow_command_sub], False):
+                integer = self.EvalToBigInt(node2)
```

Comment:
On inspection, I think this could be problematic because we are evaluating `node2` when   `eval_unsafe_arith` is OFF

----

The reason eval_unsafe_arith exists is because of this design flaw in shell, pointed out by the Fedora security team, and which I re-discovered quite a few years ago (and reported to OpenBSD  and they fixed !!!   bash still has the bug)

https://github.com/oils-for-unix/blog-code/tree/main/crazy-old-bug

----

I am thinking we might want something like

rename:

match.LooksLikeInteger -> match.LooksLikeDecimalInt   (also search for LOOKS_LIKE_INTEGER)

and then add

```
match.LooksLikeHexInt
match.LooksLikeOctInt
match.LooksLikeShellInt  # arbitrary base
```

with a regular expression.

----

The idea is to detect `0123` but not `0 + 1`     , which I think is the bug


I am not sure if this is the best approach but it seems consistent with what we already do


and it avoids the issue of "try parsing" and then evaluate

I guess it also avoids invoking the arith parser many times?

The regular language checks should be fast.   (they don't allocate!)

(Side note: the way the code is laid out with LOOKS_LIKE_INTEGER etc. is a little tortured, but I think you can just follow the pattern.  The reason it looks weird is because we have this whole `fastlex` Python extension ... in theory we can get rid of it, now that Oils is translated to C++.    There can be just 2 implementations/translations of the lexer, not 2.5 or 3!)